Grafana Observational graphs for Pipelines Controller Monitoring

ramessesii2 opened the pull request:
### The goal is to get a better understanding of how our components behave under load so that we can make appropriate tunings/adjustments.

### The following metrics are being queried for the dashboard:-

    * tekton_pipelines_controller_running_pipelineruns_count -
    Number of pipelineruns executing currently
    * tekton_pipelines_controller_running_taskruns_count -
    Number of taskruns executing currentl
    * tekton_pipelines_controller_taskruns_pod_latency
    Scheduling latency for the taskruns pods
    * tekton_pipelines_controller_workqueue_depth -
    The workqueue for the Tekton Pipelines controller
    * Pipelines controller restarts
    * tekton_pipelines_controller_client_resutls -
     Total number of K8s API requests made by Pipelines Client


### Preview
![image](https://github.com/openshift-pipelines/pipeline-service/assets/54385449/fd950412-4c94-4737-b190-ff205a6f7025)

gabemontero commented on operator/gitops/argocd/grafana/dashboards/pipeline-service-dashboard.json:
```diff
@@ -387,6 +388,420 @@
+      "targets": [
+        {
+          "exemplar": true,
+          "expr": "avg(tekton_pipelines_controller_workqueue_depth{name=\"github.com.tektoncd.pipeline.pkg.reconciler.pipelinerun.Reconciler-consumer\"})",
```
We should not just do `Reconciler-consumer`

Looking at prod right now from the openshift monitoring console, I see real numbers for `Reconciler-fast` for example.

Either avg across all the reconciler queues, i.e. lose the name qualifier, or create a chart for each one.  

My preference is avg across all queues, but chime in here with your thoughts around this (including anything you may have put in the google doc that I missed).

ramessesii2 replied:
TBH, one of my efforts has been to understand watching which components will be meaning for us. I don't usually get 0/1 binary answers to these queries. Your suggestion sounds good to me, Gabe 😃 . And if in future we find a need of individual reconciler queues, we can always reiterate over the JSON.

gabemontero commented:
by the way @ramessesii2 I have confirmed that openshift-pipelines 1.11.0 has the quota/node throttling metrics I recently added upstream:

```
gmontero ~/bin $ curl http://127.0.0.1:9090/metrics | grep tekton_pipelines_controller_running_taskruns_throttled
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0# HELP tekton_pipelines_controller_running_taskruns_throttled_by_node_count Number of taskruns executing currently, but whose underlying Pods or Containers are suspended by k8s because of Node level constraints. Such suspensions can occur as part of initial scheduling of the Pod, or scheduling of any of the subsequent Container(s) in the Pod after the first Container is started
# TYPE tekton_pipelines_controller_running_taskruns_throttled_by_node_count gauge
tekton_pipelines_controller_running_taskruns_throttled_by_node_count 0
# HELP tekton_pipelines_controller_running_taskruns_throttled_by_quota_count Number of taskruns executing currently, but whose underlying Pods or Containers are suspended by k8s because of defined ResourceQuotas.  Such suspensions can occur as part of initial scheduling of the Pod, or scheduling of any of the subsequent Container(s) in the Pod after the first Container is started
# TYPE tekton_pipelines_controller_running_taskruns_throttled_by_quota_count gauge
tekton_pipelines_controller_running_taskruns_throttled_by_quota_count 0
```

It is not in 1.10.4.

Assuming when our move to 1.11 happens, I think it makes sense to add those metrics along with the ones you have here.

Also note my findings around the latency metric i.e. https://redhat-internal.slack.com/archives/C032EJ007C0/p1686932611229559 ... I'm beginning to think we don't use that metrics, and just use the exporter's `pipelinerun_duration_scheduled_seconds_[bucket, sum, count]` metric once you get the dashboard working with the exporter.

ramessesii2 commented:
> I'm beginning to think we don't use that metrics, and just use the exporter's pipelinerun_duration_scheduled_seconds_[bucket, sum, count] metric once you get the dashboard working with the exporter.

Sure Gabe, I'm working on fixing the `Exporter metrics in Grafana`. I'll keep this note.

adambkaplan commented:
Small nit - commit message appears to be missing a "title." The general recommendation is that the first line of a commit message should be fewer than 50 characters.

ramessesii2 commented:
I've addressed your reviews @gabemontero @adambkaplan.
Although, I've made the final changes and resolved any conflicts that came after @prietyc123's [PR](https://github.com/openshift-pipelines/pipeline-service/pull/696) is merged, I can't say for sure that it's gonna look the same as the screenshot in the PR's description. This is because Prity's changes aren't yet reflected in the Grafana in Staging envs.

gabemontero commented:
@ramessesii2 and I talked during scrum.  He will take on the next infra-deployment bump to dev/stage to pull in @prietyc123 's change into stage.

And also, once we merge @adambkaplan 's inner dev loop work with Grafana, that will be an option as well.

Ideally yeah let's get that dev/stage or inner loop thing ready so @ramessesii2 can verify his rebase on top of @prietyc123 's change.

ramessesii2 commented:
Thanks Gabe.
The infra-bump PR is up here - https://github.com/redhat-appstudio/infra-deployments/pull/2033.
I'll rebase once it's merged.

gabemontero commented:
ok @ramessesii2 both @adambkaplan inner loop and https://github.com/redhat-appstudio/infra-deployments/pull/2033 are merged, and I see @prietyc123 's new PaC charts on staging.

Go ahead and rebase when you are next back at work.

I'll also let you drive if we bump to prod with the https://github.com/redhat-appstudio/infra-deployments/pull/2033 changes or if you wait until your changes are in staging, and then we bump that to prod after vetting there.

@adambkaplan @Roming22 FYI

ramessesii2 commented:
This PR has been rebased. I've also updated the preview image. Since @gabemontero & @adambkaplan are on PTO this week. I'll ask @Roming22 to do me the favor of merging this PR if everything looks good to you.

ramessesii2 commented:
I'm seeing the same error in #703 with deploy-cluster and destroy-cluster `Pipelines as Code CI / pipeline-service-upgrade-test` failing. To me it seems unrelated although I've not been able to look at the logs.

ramessesii2 commented:
/test Pipelines as Code CI / pipeline-service-upgrade-test
